keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,802 @@
+      #     this field is required
+      name: super-important-kube-system-authorizer
+      # The duration to cache 'authorized' responses from the webhook
+      # authorizer.
+      # Same as setting `--authorization-webhook-cache-authorized-ttl` flag
+      # Default: 5m0s
+      authorizedTTL: 30s
+      # The duration to cache 'authorized' responses from the webhook
+      # authorizer.
+      # Same as setting `--authorization-webhook-cache-unauthorized-ttl` flag
+      # Default: 30s
+      unauthorizedTTL: 30s
+      # Timeout for the webhook request
+      # Maximum allowed is 30s.
+      # Required, with no default.
+      timeout: 3s
+      # The API version of the authorization.k8s.io SubjectAccessReview to
+      # send to and expect from the webhook.
+      # Same as setting `--authorization-webhook-version` flag
+      # Required, with no default
+      # Valid values: v1beta1, v1
+      subjectAccessReviewVersion: v1
+      # Controls the authorization decision when a webhook request fails to
+      # complete or returns a malformed response or errors evaluating
+      # matchConditions.
+      # Valid values:
+      #   - NoOpinion: continue to subsequent authorizers to see if one of
+      #     them allows the request
+      #   - Deny: reject the request without consulting subsequent authorizers
+      # Default: NoOpinion
+      failurePolicy: Deny
+      connectionInfo:
+        # Controls how the webhook should communicate with the server.
+        # Valid values:
+        # - KubeConfig: use the file specified in kubeConfigFile to locate the
+        #   server.
+        # - InClusterConfig: use the in-cluster configuration to call the
+        #   SubjectAccessReview API hosted by kube-apiserver. This mode is not
+        #   allowed for kube-apiserver.
+        type: KubeConfig
+        # Path to KubeConfigFile for connection info
+        # Required, if connectionInfo.Type is KubeConfig
+        kubeConfigFile: /kube-system-authz-webhook.yaml
+        # matchConditions is a list of conditions that must be met for a request to be sent to this
+        # webhook. An empty list of matchConditions matches all requests.
+        # There are a maximum of 64 match conditions allowed.
+        #
+        # The exact matching logic is (in order):
+        #   1. If ANY matchCondition evaluates to FALSE, the webhook is skipped.
+        #   2. If ALL matchConditions evaluate to TRUE, the webhook is called.
+        #   3. If any matchCondition evaluates to an error (but none are FALSE):
+        #      - If failurePolicy=Deny, reject the request
+        #      - If failurePolicy=NoOpinion, the error is ignored and the webhook is skipped
+      matchConditions:
+      # expression represents the expression which will be evaluated by CEL. Must evaluate to bool.
+      # CEL expressions have access to the contents of the SubjectAccessReview
+      # in the version specified by subjectAccessReviewVersion in the request variable.
```

Comment:
Do we really want to tie the CEL expression input to the version of the webhook?

Reply:
We need to support both versions for the time being. I don't see a path forward without tying to the version unless we internally convert between the versions, which would make it more complicated and less transparent to users.

Reply:
I think it would be better to convert internally to v1 always because it keeps all of our CEL examples consistent and valid regardless of how the webhook works.  Doubling the CEL surface area would be a mistake.

Reply:
Doesn't it look odd in terms of the config if we allow both versions for the webhook request, but then have CEL filters apply to only v1 after internal conversions?

Reply:
Think about this way: does it really provide value to the end user to have to learn _two_ slightly different ways to write CEL filters?  The user configuring the API server may not be the webhook author.  IMO we are conflating two unrelated things here - one is about how the API server makes decisions, and the other is about how we ask an external webhook how to make decisions.  These are not the same, and CEL is net new and has no reason to support a beta format.

Reply:
> The user configuring the API server may not be the webhook author.

Correct.

> These are not the same, and CEL is net new and has no reason to support a beta format.

Makes sense.